tc/core/compiler-inl.h

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "tc/core/compiler.h"
+
+#include <sstream>
+#include <string>
+
+#include "tc/core/flags.h"
+#include "tc/core/halide_utils.h"
+#include "tc/core/tensor.h"
+#include "tc/lang/canonicalize.h"
+
+namespace tc {
+namespace {
+void checkInputsCompliant(
+    const tc2halide::HalideComponents& halideComponents,
+    const std::vector<const DLConstTensor*>& inputsInfo) {
+  if (inputsInfo.size() != halideComponents.inputs.size()) {
+    throw lang::ErrorReport(halideComponents.getDef())
+        << "expected " << halideComponents.inputs.size() << " inputs but found "
+        << inputsInfo.size();
+  }
+  for (size_t i = 0; i < inputsInfo.size(); ++i) {
+    auto dltype = inputsInfo[i]->dtype;
+    auto htype = halideComponents.inputs[i].type();
+    // we have three type representations here: (1) halide Type (2) DLTensor
+    // type, and (3) the token representing the type in the frontend (e.g.
+    // TK_FLOAT) we need to translate to (3) to report user facing errors
+    auto dltype =
+        lang::TypeInfo(lang::TypeInfo::Code(dltype.code), dltype.bits)
+            .toScalarToken();
+    auto htype =
+        lang::TypeInfo(lang::TypeInfo::Code(htype.code()), htype.bits())
+            .toScalarToken();
+    if (dltype != htype) {
+      throw lang::ErrorReport(halideComponents.getDef().params()[i])
+          << "expected type " << lang::kindToString(htype) << " but found "
+          << lang::kindToString(dltype);
+    }
+    int hdim = halideComponents.inputs[i].dimensions();
+    int dldim = inputsInfo[i]->ndim;
+    if (dldim != hdim) {
+      throw lang::ErrorReport(halideComponents.getDef().params()[i])
+          << "expected a tensor with " << hdim << " dimensions but found "
+          << dldim << " dimensions.";
+    }
+  }
+}
+} // namespace
+template <typename Backend>
+std::unique_ptr<typename Backend::ExecutorType> compile(
+    const std::string& tc,
+    const std::string& entryPoint,
+    const std::vector<const DLConstTensor*>& inputs,
+    /* TODO: in the future also pass outputs for stride and alignment info */
+    const typename Backend::MappingOptionsType& options) {
+  auto parsedTcs = detail::parse(tc);
+  CHECK_EQ(1, parsedTcs.count(entryPoint))
```

Comment:
do you mind to start using ENFORCE instead of a check here and in other places? Check crashes the process. And this can be a problem going forward with productionization (if we put this code in inference). 
Feel free to borrow CAFFE_ENFORCE code :) 

Regarding CHECK_EQ it should be preferred over CHECK, agree with Nicolas. As it automatically prints values. For this reason we have ENFORCE_EQ :)

Reply:
Please see #351, the idea we floated was to use exceptions everywhere, will happen globally over the whole codebase when it is addressed.
Do you see a value in maintaining 2 sets of macros?
Care to comment and add suggestions to #351?

Thanks!